sample.go

```diff
@@ -0,0 +1,215 @@
+			indentVal(w, val, indentStr)
+		} else {
+			if f.Example != nil {
+				val = f.Example
+			} else {
+				val = sampleValue(f.Type)
+			}
+			fmt.Fprintf(w, "# %s:", f.name)
+			indentVal(w, val, indentStr+"# ")
+		}
+	}
+	return nil
+}
+
+const textWidth = 80
+
+func writeSampleDescription(w io.Writer, f Attr, indent string) {
+	previousText := false
+
+	// section marks the start of a new section of the comment;
+	// sections are separated with empty lines.
+	section := func() {
+		if previousText {
+			fmt.Fprintf(w, "%s\n", strings.TrimRightFunc(indent, unicode.IsSpace))
+		}
+		previousText = true
+	}
+
+	descr := strings.TrimSpace(f.Description)
+	if descr != "" {
+		section()
+		if !strings.HasSuffix(descr, ".") {
+			descr += "."
+		}
+		doc.ToText(w, descr, indent, "    ", textWidth-len(indent))
+	}
+	vars := make([]string, 0, len(f.EnvVars)+1)
+	if f.EnvVar != "" {
+		vars = append(vars, "$"+f.EnvVar)
+	}
+	for _, v := range f.EnvVars {
+		vars = append(vars, "$"+v)
+	}
+	if len(vars) > 0 {
+		section()
+		fmt.Fprintf(w, "%sDefault value taken from %s.\n", indent, wordyList(vars))
+	}
+	attrText := ""
+	switch {
+	case f.Secret && f.Immutable:
+		attrText = "immutable and considered secret"
+	case f.Secret:
+		attrText = "considered secret"
+	case f.Immutable:
+		attrText = "immutable"
+	}
+	if attrText != "" {
+		section()
+		fmt.Fprintf(w, "%sThis attribute is %s.\n", indent, attrText)
+	}
+	section()
+}
+
+func emptyLine(w io.Writer, indent string) {
+	fmt.Fprintf(w, "%s\n", strings.TrimRightFunc(indent, unicode.IsSpace))
+}
+
+func wordyList(words []string) string {
+	if len(words) == 0 {
+		return ""
+	}
+	if len(words) == 1 {
+		return words[0]
+	}
+	return strings.Join(words[0:len(words)-1], ", ") + " or " + words[len(words)-1]
+}
+
+var groupPriority = map[Group]int{
+	ProviderGroup: 3,
+	AccountGroup:  2,
+	EnvironGroup:  1,
+}
+
+type attrWithName struct {
+	name string
+	Attr
+}
+
+type fieldsByGroup []attrWithName
+
+func (f fieldsByGroup) Len() int {
+	return len(f)
+}
+
+func (f fieldsByGroup) Swap(i, j int) {
+	f[i], f[j] = f[j], f[i]
+}
+
+func (f fieldsByGroup) Less(i0, i1 int) bool {
+	f0, f1 := &f[i0], &f[i1]
+	pri0, pri1 := groupPriority[f0.Group], groupPriority[f1.Group]
+	if pri0 != pri1 {
+		return pri0 > pri1
+	}
+	return f0.name < f1.name
+}
+
+var (
+	space = []byte(" ")
+	nl    = []byte("\n")
+)
+
+// indentVal writes the given YAML-formatted value x to w and prefixing
+// the second and subsequent lines with the given ident.
+func indentVal(w io.Writer, x interface{}, indentStr string) {
+	data, err := yaml.Marshal(x)
+	if err != nil {
+		panic(fmt.Errorf("cannot marshal YAML", err))
+	}
+	if len(data) == 0 {
+		panic("YAML cannot marshal to empty string")
+	}
+	indent := []byte(indentStr + "  ")
+	if canUseSameLine(x) {
+		w.Write(space)
+	} else {
+		w.Write(nl)
+		w.Write(indent)
+	}
+	data = bytes.TrimSuffix(data, nl)
+	lines := bytes.Split(data, nl)
+	for i, line := range lines {
```

Comment:
any particular reason not to merge this and the above line together?
`for i, line := bytes.Split(data, nl)` No particular reason to do so either.